pennylane/transforms/split_non_commuting.py

```diff
@@ -0,0 +1,113 @@
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains the tape transform that splits non-commuting terms
+"""
+# pylint: disable=protected-access
+import pennylane as qml
+import numpy as np
+
+
+def split_non_commuting(tape):
+    r"""
+    Splits a tape measuring non-commuting observables into groups of commuting observables.
+
+    Args:
+        tape (.QuantumTape): the tape used when calculating the expectation value
+            of the Hamiltonian
+
+    Returns:
+        tuple[list[.QuantumTape], function]: Returns a tuple containing a list of
+        quantum tapes to be evaluated, and a function to be applied to these
+        tape executions to restore the ordering of the inputs.
+
+    **Example**
+
+    We can create a tape with non-commuting observables:
+
+    .. code-block:: python3
+
+        import pennylane as qml
+        with qml.tape.QuantumTape() as tape:
+            qml.expval(qml.PauliZ(0))
+            qml.expval(qml.PauliY(0))
+        tapes, processing_fn = qml.transforms.split_non_commuting(tape)
+
+    Now ``tapes`` is a list of two tapes, each for one of the non-commuting terms:
+
+    >>> [t.observables for t in tapes]
+    [[expval(PauliZ(wires=[0]))], [expval(PauliY(wires=[0]))]]
+
+    The processing function becomes important when creating the commuting groups as the order of the inputs has been modified:
+
+    .. code-block:: python3
+
+        with qml.tape.QuantumTape() as tape:
+            qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+            qml.expval(qml.PauliX(0) @ qml.PauliX(1))
+            qml.expval(qml.PauliZ(0))
+            qml.expval(qml.PauliX(0))
+        tapes, processing_fn = qml.transforms.split_non_commuting(tape)
+
+    In this example, the groupings are ``group_coeffs = [[0,2], [1,3]]`` and processing_fn makes sure that the final output is of the same shape and ordering:
+
+    >>> processing_fn(tapes)
+    tensor([tensor(expval(PauliZ(wires=[0]) @ PauliZ(wires=[1])), dtype=object, requires_grad=True),
+        tensor(expval(PauliX(wires=[0]) @ PauliX(wires=[1])), dtype=object, requires_grad=True),
+        tensor(expval(PauliZ(wires=[0])), dtype=object, requires_grad=True),
+        tensor(expval(PauliX(wires=[0])), dtype=object, requires_grad=True)],
+       dtype=object, requires_grad=True)
+
+    """
+    # Even though we are currently not supporting probs and samples, we still provide the option here
+    # as it works in some cases.
+    obs_fn = {qml.measurements.Expectation: qml.expval,
+              qml.measurements.Variance: qml.var,
+              qml.measurements.Sample: qml.sample,
+              qml.measurements.Probability: qml.probs}
```

Comment:
What happens when/if it doesn't work?

Reply:
Actually I am not sure if there are any actual working examples. I tried something like
```python
@qml.qnode(dev)
def circuit():
    return [qml.probs(wires=0), qml.expval(qml.PauliX(0))]
```
but then it seems it is automatically assuming to measure the probabilities in the X basis (I thought that it would always do the computational basis and thus this being a non-commuting measurement). When I put a specific basis and/or add another operator like in the following example
```python
@qml.qnode(dev)
def circuit():
    return [qml.probs(op=qml.PauliX(0)), qml.expval(qml.PauliX(0)), qml.expval(qml.PauliZ(0))]
circuit()
```
then I obtain
```
File ~/Qottmann/Xanadu/pennylane/pennylane/devices/default_qubit.py:184, in <listcomp>(.0)
    [183](file:///home/qottmann/Qottmann/Xanadu/pennylane/pennylane/devices/default_qubit.py?line=182) try:
--> [184](file:///home/qottmann/Qottmann/Xanadu/pennylane/pennylane/devices/default_qubit.py?line=183)     mapped_wires = [self.wire_map[w] for w in wires]
    [185](file:///home/qottmann/Qottmann/Xanadu/pennylane/pennylane/devices/default_qubit.py?line=184) except KeyError as e:

KeyError: PauliZ(wires=[0])

The above exception was the direct cause of the following exception:
(...)
WireError: Did not find some of the wires (PauliZ(wires=[0]),) on device with wires (0,)`
```